cputest/psq.cpp

```diff
@@ -0,0 +1,807 @@
+
+// quantize types
+enum EQuantizeType
+{
+	QUANTIZE_FLOAT = 0,
+	QUANTIZE_U8    = 4,
+	QUANTIZE_U16   = 5,
+	QUANTIZE_S8    = 6,
+	QUANTIZE_S16   = 7,
+};
+
+union UGQR
+{
+	BitField< 0, 3, u32> st_type;
+	BitField< 8, 6, u32> st_scale;
+	BitField<16, 3, u32> ld_type;
+	BitField<24, 6, u32> ld_scale;
+
+	u32 hex;
+};
+
+// dequantize table
+const float m_dequantizeTable[] =
+{
+	1.0 / (1ULL <<  0), 1.0 / (1ULL <<  1), 1.0 / (1ULL <<  2), 1.0 / (1ULL <<  3),
+	1.0 / (1ULL <<  4), 1.0 / (1ULL <<  5), 1.0 / (1ULL <<  6), 1.0 / (1ULL <<  7),
+	1.0 / (1ULL <<  8), 1.0 / (1ULL <<  9), 1.0 / (1ULL << 10), 1.0 / (1ULL << 11),
+	1.0 / (1ULL << 12), 1.0 / (1ULL << 13), 1.0 / (1ULL << 14), 1.0 / (1ULL << 15),
+	1.0 / (1ULL << 16), 1.0 / (1ULL << 17), 1.0 / (1ULL << 18), 1.0 / (1ULL << 19),
+	1.0 / (1ULL << 20), 1.0 / (1ULL << 21), 1.0 / (1ULL << 22), 1.0 / (1ULL << 23),
+	1.0 / (1ULL << 24), 1.0 / (1ULL << 25), 1.0 / (1ULL << 26), 1.0 / (1ULL << 27),
+	1.0 / (1ULL << 28), 1.0 / (1ULL << 29), 1.0 / (1ULL << 30), 1.0 / (1ULL << 31),
+	(1ULL << 32), (1ULL << 31), (1ULL << 30), (1ULL << 29),
+	(1ULL << 28), (1ULL << 27), (1ULL << 26), (1ULL << 25),
+	(1ULL << 24), (1ULL << 23), (1ULL << 22), (1ULL << 21),
+	(1ULL << 20), (1ULL << 19), (1ULL << 18), (1ULL << 17),
+	(1ULL << 16), (1ULL << 15), (1ULL << 14), (1ULL << 13),
+	(1ULL << 12), (1ULL << 11), (1ULL << 10), (1ULL <<  9),
+	(1ULL <<  8), (1ULL <<  7), (1ULL <<  6), (1ULL <<  5),
+	(1ULL <<  4), (1ULL <<  3), (1ULL <<  2), (1ULL <<  1),
+};
+
+u32 GenerateGQR(u8 st_type, u8 st_scale, u8 ld_type, u8 ld_scale)
+{
+	UGQR gqr;
+	gqr.st_type = st_type;
+	gqr.st_scale = st_scale;
+	gqr.ld_type = ld_type;
+	gqr.ld_scale = ld_scale;
+	return gqr.hex;
+}
+
+template<u8 gqr>
+void SetGQR(u32 value)
+{
+	static_assert(gqr < 8, "GQR being set must be below 8!");
+	asm("mtspr %[spr], %0" :: "r" (value), [spr] "i" (912 + gqr));
+}
+
+static void psqlSingleU8Test()
+{
+	START_TEST();
+	for (u8 scale = 0; scale <= 0b111111; ++scale)
+	{
+		u8 values[256];
+		float results[256];
+
+		for (u32 i = 0; i < 256; ++i)
+			values[i] = i;
+
+		// Better pray your basic float multiplication is correct
+		for (u32 i = 0; i < 256; ++i)
+			results[i] = m_dequantizeTable[scale] * (float)i;
+
+		for (size_t i = 0; i < sizeof(values) / sizeof(values[0]); i++)
+		{
+			float result = 0.0;
+
+			SetGQR<0>(GenerateGQR(0, 0, QUANTIZE_U8, scale));
+			asm("psq_l %[result], 0(%[loc]), 1, 0"
+				: [result] "=f" (result)
+				: [loc] "r" ((u32)&values[i]));
+			DO_TEST(result == results[i], "psq_l-u8(%d)(%d):\n"
+							     "\tgot %08x(%f)\n"
+							     "\texpected %08x(%f)", values[i], scale, result, result, results[i], results[i]);
+		}
+	}
+
+	END_TEST();
+}
+
+static void psqlSingleS8Test()
+{
+	START_TEST();
+	for (u8 scale = 0; scale <= 0b111111; ++scale)
+	{
+		s8 values[256];
+		float results[256];
+
+		for (s32 i = 0; i < 256; ++i)
+			values[i] = i - 128;
+
+		// Better pray your basic float multiplication is correct
```

Comment:
Useless comment: is there any actual danger from using this? Is it a TODO for stuff which actually needs to be verified?

Please explain better what you were thinking when writing this comment.